diff --git a/QAppConfig/QAppLauncher/skins/default/styleSheet.qss b/QAppConfig/QAppLauncher/skins/default/styleSheet.qss
index 95e505f..d0f9a77 100644
--- a/QAppConfig/QAppLauncher/skins/default/styleSheet.qss
+++ b/QAppConfig/QAppLauncher/skins/default/styleSheet.qss
@@ -6,7 +6,7 @@
 
 QAppLaunchControl {
 	background-color:rgb(0, 0, 0);
-    background-image: url(QAppConfig/QAppLauncher/skins/default/bg.png);
+    background-image: url(/opt/QAppConfig/QAppLauncher/skins/default/bg.png);
     background-repeat: no-repeat;
     background-position: center;
 }
