diff -rup linux-nanopi-v4.1.y.orig/net/wireless/reg.c linux-nanopi-v4.1.y/net/wireless/reg.c
--- linux-nanopi-v4.1.y.orig/net/wireless/reg.c	2015-08-12 01:38:07.000000000 -0700
+++ linux-nanopi-v4.1.y/net/wireless/reg.c	2015-09-06 01:24:40.495808769 -0700
@@ -136,7 +136,7 @@ static spinlock_t reg_indoor_lock;
 static u32 reg_is_indoor_portid;
 
 /* Max number of consecutive attempts to communicate with CRDA  */
-#define REG_MAX_CRDA_TIMEOUTS 10
+#define REG_MAX_CRDA_TIMEOUTS 0
 
 static u32 reg_crda_timeouts;
 
